Add otel.sdk.component.shutdown self-observability event (#3723)
open-telemetry/semantic-conventions: docs/otel/sdk-events.md

| `otlp_http_span_exporter` | OTLP span exporter over HTTP with protobuf serialization | ![Development](https://img.shields.io/badge/-development-blue) |
| `periodic_metric_reader` | The built-in SDK periodically exporting metric reader | ![Development](https://img.shields.io/badge/-development-blue) |
| `prometheus_http_text_metric_exporter` | Prometheus metric exporter over HTTP with the default text-based format | ![Development](https://img.shields.io/badge/-development-blue) |
| `simple_log_processor` | The built-in SDK simple log record processor | ![Development](https://img.shields.io/badge/-development-blue) |
| `simple_span_processor` | The built-in SDK simple span processor | ![Development](https://img.shields.io/badge/-development-blue) |
| `tracer_provider` | The built-in SDK TracerProvider | ![Development](https://img.shields.io/badge/-development-blue) |
| `zipkin_http_span_exporter` | Zipkin span exporter over HTTP | ![Development](https://img.shields.io/badge/-development-blue) |

<!-- prettier-ignore-end -->
<!-- END AUTOGENERATED TEXT -->
<!-- endsemconv -->

> The component being shut down is identified by the existing
> [`otel.component.*`](/docs/registry/attributes/otel.md#otel-component-attributes) attributes,
> not encoded in the event name. This keeps the event name stable and reusable
> as the set of emitting components grows.
> 

Reviewer: The first sentence restates the note at the top of this same section ("The component that shut down is identified by `otel.component.type` and `otel.component.name`"), and the second is design rationale about future extensibility - the kind of thing that reads well in the PR description but not in the convention itself.

```suggestion
```

<!--- Hugo front matter used to generate the website version of this page:
linkTitle: SDK Events
--->

# Semantic conventions for OpenTelemetry SDK events

**Status**: [Development][DocumentStatus]

Self-observability events emitted by OpenTelemetry SDK components.

These events complement SDK self-observability metrics by providing per-instance
confirmation of lifecycle transitions, such as whether shutdown completed cleanly.

> [!NOTE]

> > Events are in-development and not yet available in some languages. Check the [spec-compliance matrix](https://github.com/open-telemetry/opentelemetry-specification/blob/main/spec-compliance-matrix.md#logs) to see the implementation status in the corresponding language.

Reviewer: `CONTRIBUTING.md` requires links to the specification repository to point to a tag matching the version in `README.md` (currently `v1.59.0`), and explicitly **not** to `main`. The equivalent note in `docs/feature-flags/feature-flags-events.md` already uses the pinned form.

```suggestion
> Events are in-development and not yet available in some languages. Check the [spec-compliance matrix](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.59.0/spec-compliance-matrix.md#logs) to see the implementation status in the corresponding language.
```

exporters, as well as custom implementations of any of these categories. See
the well-known values of
[`otel.component.type`](/docs/registry/attributes/otel.md) for the full list of
built-in types.

Providers (e.g. `TracerProvider`, `LoggerProvider`, `MeterProvider`) are not
expected to emit this event. A provider's shutdown is already covered by the
events emitted by its child components, and a higher-level provider event would
carry no information beyond what those child events already convey. A future
revision of this convention may define provider-level shutdown events if a need
emerges.

**Emission rules.**

At most one `otel.sdk.component.shutdown` event is emitted per component
instance per process lifetime, reporting the result of that one shutdown.

If the process terminates without invoking the component's shutdown operation
(e.g. a crash, `SIGKILL`, container eviction, or immediate process termination),
no event is expected. Consumers MUST NOT interpret the absence of this event as
evidence of a clean shutdown.

The event timestamp SHOULD reflect the time at which the shutdown attempt ended
(whether by completing, failing, or being abandoned). Combined with

> `otel.component.shutdown.duration`, the start of the shutdown attempt can be

Reviewer 2: tiny clarification:

```suggestion
`otel.component.shutdown.duration` attribute, the start of the shutdown attempt can be
```

(or a link to attribute in the registry, they are anchored)

I didn't realize there is an attribute when started reading this and thought it's a metric.

reconstructed as `timestamp - duration`.

**Severity.**

Implementations MUST set the log record severity to `WARN`
(`SeverityNumber` = `13`) when `otel.component.shutdown.result` is not

> `success`, and to `INFO` (`SeverityNumber` = `9`) otherwise.

Reviewer 2: why not DEBUG? 

E.g. we'd eventually have some form of SDK shutdown event - that one is important enough to be INFO, the components at DEBUG would just provide verbose details.

Author: resolving based on latest iteration which makes this a provider level one, and INFO.

**Nesting.**

When an SDK component invokes the shutdown of a child component inline as
part of its own shutdown (e.g. a batching processor shutting down its
exporter), the parent's `otel.component.shutdown.duration` mechanically
includes the time spent shutting down its children.

**Attributes:**

| Key | Stability | [Requirement Level](https://opentelemetry.io/docs/specs/semconv/general/attribute-requirement-level/) | Value Type | Description | Example Values |
| --- | --- | --- | --- | --- | --- |
| [`otel.component.name`](/docs/registry/attributes/otel.md) | ![Development](https://img.shields.io/badge/-development-blue) | `Required` | string | A name uniquely identifying the instance of the OpenTelemetry component within its containing SDK instance. [1] | `otlp_grpc_span_exporter/0`; `custom-name` |

> | [`otel.component.shutdown.result`](/docs/registry/attributes/otel.md) | ![Development](https://img.shields.io/badge/-development-blue) | `Required` | string | The result of an OpenTelemetry SDK component shutdown attempt. [2] | `success`; `failed`; `timed_out` |

Reviewer 2: pls include error.type - a canonical way to record error class across conventions. It's presence already implies failure. 

Timeout is considered to be a failure by default (we can argue).

Assuming we add `error.type`, I think we should remove this attribute - I can see it being useful if we have more statuses or need to separate timeout from error, but it does not seem as important here as for e.g. http calls.

Author: agree, switched to this in latest iteration.
